### DISCUSSION TITLE
[Feature Request] Ability to duplicate charts

### DISCUSSION THREAD

**Discussion by LuPan2015** [apache/superset#19254] [Ideas]
Currently, we can only create a chart from scratch, instead of modifying it based on an existing chart。

**Comment by zhaoyongjie** [comment]
> Currently, we can only create a chart from scratch, instead of modifying it based on an existing chart。

I don't understand this issue, Could you detail it?

**Comment by LuPan2015** [comment]
> > Currently, we can only create a chart from scratch, instead of modifying it based on an existing chart。
> 
> I don't understand this issue, Could you detail it?

Similar to copy action in role. As shown below

![image](https://user-images.githubusercontent.com/10804128/158941633-8cb4cac7-968a-4f30-b5b7-2203add78712.png)

This way I can quickly create a new chart based on an existing chart。

**Comment by srinify** [comment]
Hey @LuPan2015 you're correct that Duplicating / Copying charts doesn't exist at the moment as a feature.

Two workarounds:

1. Export and Import the chart and give it a new name.
OR 

2. Open the chart in Explore and when you click Save, save it as a new chart with a new name (don't overwrite).

I'm converting this thread to a Github Discussion, since we use Github Issues to focus on bugs in existing functionality :)

**Comment by levi-applied** [comment]
Thanks for these workarounds, I'm using them now but I should say that not having the ability to copy charts has annoyed me for a very long time. So much so that I came to file a feature request (and found this thread that already requested the same)

**Comment by bkozura** [comment]
I second this. It would be great to have:

1. A "save as" option in the chart editor like we do in the dashboard screen. 

      current chart options:
      <img width="155" alt="current chart options" src="https://github.com/apache/superset/assets/97912474/de9a496f-aa3c-48ee-a1f7-b062beb0435f">

      current dashboard options:
    <img width="213" alt="current dashboard options" src="https://github.com/apache/superset/assets/97912474/f0ac1753-a1b3-4fc9-9bb9-c245c49f94e3">



2. A "Duplicate chart" feature within the dashboard that creates a copy of the original 

      current chart options within dashboard:
      <img width="497" alt="current chart options within dashboard" src="https://github.com/apache/superset/assets/97912474/b8788991-4fc5-4a9a-bb9d-14f88b19f986">

**Comment by jplahn** [comment]
The workarounds certainly work fine, but I'd love a first class option to duplicate a chart.

**Comment by Sejmou** [comment]
I understand that maintaning this project is a lot of work and hence this feature hasn't been of high priority. However, this really feels like a pretty essential feature and shouldn't require workarounds. Are there any plans to introduce it any time soon?

I would enjoy the process of creating charts much more (and be a lot quicker) if I could just use charts I had already built before as a starting point 😅

**Comment by Sejmou** [comment]
As I had to copy quite a lot of charts recently, I figured I might as well hack something together to automate the most annoying steps. 

This Python script creates a copy of any ZIP archive for an exported chart, swapping the chart's UUID w/ a freshly generated one so that the chart copy can be imported into Superset without errors.

```python
import zipfile
import os
import uuid
import shutil
import yaml
import argparse


def unzip_archive(zip_file, extract_to):
    """Unzip the archive to the specified directory."""
    with zipfile.ZipFile(zip_file, "r") as zip_ref:
        zip_ref.extractall(extract_to)


def edit_yaml_uuid(yaml_file, new_uuid):
    """Edit the YAML file, replacing the uuid property with a new UUID."""
    with open(yaml_file, "r") as file:
        content = yaml.safe_load(file)

    # Assuming the 'uuid' is a top-level property
    content["uuid"] = str(new_uuid)

    with open(yaml_file, "w") as file:
        yaml.dump(content, file)


def zip_directory(directory, output_filename):
    """Zip the directory into a zip file."""
    shutil.make_archive(output_filename, "zip", directory)


def copy_chart_zip(zip_file):
    new_uuid = uuid.uuid4()
    working_dir = f"tmp_{new_uuid}"

    # Unzip the archive
    unzip_archive(zip_file, working_dir)

    chart_parent_dir_path = None
    zip_contents = os.listdir(working_dir)
    for item in zip_contents:
        if item.startswith("chart_export"):
            chart_export_dirname = item
            chart_parent_dir_path = os.path.join(working_dir, chart_export_dirname)
            break
    if not chart_parent_dir_path:
        shutil.rmtree(working_dir)
        raise ValueError(
            "Expected ZIP archive to contain a directory starting with 'chart_export'. Are you sure this is a Superset Chart Export ZIP file?"
        )

    chart_yml_file = None
    chart_yml_dir = os.path.join(chart_parent_dir_path, "charts")
    if not os.path.exists(chart_yml_dir):
        shutil.rmtree(working_dir)
        raise ValueError(
            "No 'charts' directory found in the extracted folder. Are you sure this is a Superset Chart Export ZIP file?"
        )

    for root, dirs, files in os.walk(chart_yml_dir):
        for file in files:
            if file.endswith(".yaml"):
                chart_yml_file = os.path.join(root, file)
                break

    if chart_yml_file:
        # Edit the UUID in the .yml file
        edit_yaml_uuid(chart_yml_file, new_uuid)

        # Create a new zip archive
        base_name = os.path.splitext(zip_file)[0] + "_copy"
        zip_directory(working_dir, base_name)

        # Clean up the extracted files
        shutil.rmtree(working_dir)
        print(f"New archive created: {base_name}.zip")
    else:
        shutil.rmtree(working_dir)
        raise ValueError(
            "No .yaml file found in the 'charts' directory. Are you sure this is a Superset Chart Export ZIP file?"
        )


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "zip_file", help="The path to the Superset Chart Export ZIP file."
    )
    args = parser.parse_args()

    copy_chart_zip(args.zip_file)
```
Just copy this to a file (e.g. `create_superset_chart_zip_copy.py`) and run it with the path to some chart export ZIP you downloaded (e.g. `create_superset_chart_zip_copy.py chart_export_20240812T125715.zip`). Make sure to `pip install pyyaml`, apart from that this script shouldn't have external dependencies.

This workaround is required because if you were to just export to ZIP and import it directly back into Superset, it would overwrite the existing chart and NOT create a copy. IIUC, this is essentially equivalent to the first workaround suggested by @srinify, but changing the _UUID_ rather than the _name_ (maybe that was the intention anyway?)

**Comment by GhaziTriki** [comment]
Copying a chart is very useful when you are taking the dataset and doing some adjustments for a different scenario. Not yet in the road-map?

**Comment by GhaziTriki** [comment]
I just double checked and I think this can be considered as closed with the "Save as..." feature.

**Comment by wuyang630** [comment]
real helpful feature in practice

**Comment by arobertn** [comment]
If the "Save as..." feature exists in released versions, I haven't been able to find it. But if it does, then it should only be a few steps to implement a right-click "Duplicate" option. I hope this could be added. It is by far the most convenient way to develop live in Grafana and I sorely miss it here in Superset.

**Reply by hgeist2** [reply]
I just had the same problem. Here's how it works in a current apache superset docker image:
1. Open any chart and just click the Save-Button on the top right
2. In the dialog you can choose overwrite or save as with two radio boxes on top
3. enter a different name, save

Hope that helps

**Reply by nalcabio-tom** [reply]
Thanks @hgeist2 . This was not easy to find.
